plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -2723,13 +2890,26 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(C
             return Optional.empty();
         }
 
+        replaceTableHandle(tableHandle, newHandle);
         return Optional.of(new ConstraintApplicationResult<>(
                 newHandle,
                 newUnenforcedConstraint.transformKeys(ColumnHandle.class::cast),
                 extractionResult.remainingExpression(),
                 false));
     }
 
+    private void replaceTableHandle(DeltaLakeTableHandle tableHandle, DeltaLakeTableHandle newHandle)
+    {
+        Iterator<DeltaLakeTableHandle> tableHandleIterator = tableHandles.iterator();
+        while (tableHandleIterator.hasNext()) {
+            if (Objects.equals(tableHandleIterator.next(), tableHandle)) {
+                tableHandleIterator.remove();
+                break;
+            }
+        }
+        tableHandles.add(newHandle);
```

Comment:
I wanted to add here a `checkState` that the `tableHandle` is actually removed, but I don't have any good solution for  the test `io.trino.plugin.deltalake.TestDeltaLakeMetadata#testApplyProjection()` which would fail if i'd add the check.